runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/OkHttpEngine.kt

```diff
@@ -51,6 +51,7 @@ public class OkHttpEngine(
         val engineResponse = mapOkHttpExceptions { engineCall.executeAsync() }
 
         callContext.job.invokeOnCompletion {
+            engineCall.cancel()
```

Comment:
**correctness**: I'm not sure we want to trigger a cancellation on _every_ invocation including successful ones. This may trigger [other behaviors/paths](https://github.com/square/okhttp/blob/5f008f382d67c79a990b31d4baebbdbc24365e64/okhttp/src/main/kotlin/okhttp3/internal/connection/RealCall.kt#L126) in okhttp ([including lifecycle methods on interceptors](https://github.com/square/okhttp/blob/5f008f382d67c79a990b31d4baebbdbc24365e64/okhttp/src/main/kotlin/okhttp3/internal/connection/RealCall.kt#L138-L143)). 


What we probably want is to know if there was a response body and if it was consumed or not. If not then cancel and close otherwise we don't need to do anything. 

Alternatively we have the right hooks in place on [HttpCall](https://github.com/smithy-lang/smithy-kotlin/blob/6e221cc0f7a31622cbadd4559f22627f4d396106/runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/HttpCall.kt#L69). We could add an `internal open fun cleanup() { }` that we [override](https://github.com/smithy-lang/smithy-kotlin/blob/b866db4ed13bc1b37de99db4463a26f07a4d3737/runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/OkHttpUtils.kt#L185) to invoke `engineResponse.body.close()` on the same thread that the execution is happening on.

Reply:
Yeah that makes sense. Let me try modifying `HttpCall` first and see how that works. Thanks!

Reply:
Discussed offline and decided to try more intelligent `invokeOnCompletion` logic instead.